llama : custom attention mask + parallel decoding + no context swaps

ggerganov opened the pull request:
close: #2060 #2813
ref: #3137

Merge ETA: ~ 27 - 30 Sep

Here we will attempt to add batched (a.k.a. parallel, multi-sequence) decoding to `llama.cpp`.
For a short summary of the plan, see https://github.com/ggerganov/llama.cpp/issues/3137#issuecomment-1722306526

### Features:

- parallel decoding with common prompt
- parallel decoding with separate prompts
- parallel decoding with continuous batching
- tree-based parallel decoding
- avoided the slow context swaps when the context becomes full

### API changes:

- add batch API:
  - add `struct llama_batch`
  - add `llama_batch_init`
  - add `llama_batch_free`
  - add temporary helper `llama_batch_get_one` for easy API migration
- deprecate `llama_get_kv_cache_token_count`
- add KV cache management API:
  - `llama_kv_cache_tokens_rm`
  - `llama_kv_cache_seq_rm`
  - `llama_kv_cache_seq_cp`
  - `llama_kv_cache_seq_keep`
  - `llama_kv_cache_seq_shift`
- update decoding API:
  - deprecate `llama_eval` and `llama_eval_embd`
  - add `llama_decode`
  - add `llama_get_logits_ith`


### Demo

#### No context swaps

Previously, we had to re-evaluate the context when it becomes full and this could take a lot of time, especially on the CPU.
Now, this is avoided by correctly updating the KV cache on-the-fly:

```bash
# run on the CPU using a small context of 256 tokens
./main -m ./models/llama-7b/ggml-model-q4_0.gguf -p "I believe the meaning of life is" --ignore-eos -c 256 -n -1 -t 8 -ngl 0
```

#### Parallel decoding - basic example

[examples/batched](https://github.com/ggerganov/llama.cpp/blob/custom-attention-mask/examples/batched/batched.cpp)

The `batched` example has been extended with an argument to specify the number of sequences to generate using the given prompt. This is a good starting point for understanding the new `llama_batch` API introduced in this PR.

```bash
# the prompt is "Hello my name is" and the number of sequences that will be generated is 8
./batched ./models/llama-7b-v2/ggml-model-f16.gguf "Hello my name is" 8
```

<details>
<summary>Generated results</summary>

```
main: n_len = 32, n_ctx = 2048, n_parallel = 8, n_kv_req = 221

 Hello my name is

main: generating 8 sequences ...

main: stream 0 finished
main: stream 1 finished
main: stream 2 finished
main: stream 3 finished
main: stream 4 finished
main: stream 5 finished
main: stream 6 finished
main: stream 7 finished

sequence 0:

Hello my name is Sharon. I am looking for a caregiver job in Austin, Texas. I'm a really good candidate you'll

sequence 1:

Hello my name is Renee. I am a 20 year old female who has been working with children for several years. I have experience working

sequence 2:

Hello my name is Diana. I am 25 years old and I am from the Philippines. I have been working as a nanny for 

sequence 3:

Hello my name is Cathy and I'm from the UK. I'm 26 years old and I'm a student. I

sequence 4:

Hello my name is Chelsea. I'm 27 years old and live in Florida (United States). I'm looking for someone

sequence 5:

Hello my name is Katherine, I'm 27 years old and I'm from the UK. I'm a very friendly, out

sequence 6:

Hello my name is Tiffany. I am a 21 year old student at the University of South Carolina. I am a certified nurs

sequence 7:

Hello my name is Renee, I am a mother of two beautiful children. I am a stay at home mom and I have been a nanny

main: decoded 216 tokens in 1.52 s, speed: 141.83 t/s

llama_print_timings:        load time =   489.06 ms
llama_print_timings:      sample time =     4.48 ms /   224 runs   (    0.02 ms per token, 49988.84 tokens per second)
llama_print_timings: prompt eval time =  1962.52 ms /   226 tokens (    8.68 ms per token,   115.16 tokens per second)
llama_print_timings:        eval time =     0.00 ms /     1 runs   (    0.00 ms per token,      inf tokens per second)
llama_print_timings:       total time =  2011.64 ms

ggml_metal_free: deallocating
```
</details>

#### Parallel decoding - server simulation

[examples/parallel](https://github.com/ggerganov/llama.cpp/blob/custom-attention-mask/examples/parallel/parallel.cpp)

Decoding 32 parallel streams for a total of 128 sequences with continuous batching in about ~26s. Token summary:

- System: 305 (encoded once at the start)
- Prompt: 2039 (summed over all inputs)
- Decoded: 7986 (summed over all sequences)
 
Using 7B LLaMA v2 Q8_0 model on single RTX 4080 with 16GB VRAM.

```bash
./bin/parallel -m ../models/llama-7b-v2/ggml-model-q8_0.gguf -n 128 -t 1 -ngl 100 -c 8192 -b 512 -s 1 -np 32 -ns 128 -cb
```

https://github.com/ggerganov/llama.cpp/assets/1991296/31c7c5a3-ef3e-4565-be26-7d177d73895b

For comparison, generating the same amount of sequences using a single stream (i.e. `-np 1`) takes ~125s on the same hardware with the same model.

When using the `parallel` example, you must make sure that your KV cache will have enough size to fit all parallel requests. If it is not able to fit them, then you will start seeing cache misses that slow down the processing significantly and can eventually fail the decoding if there is no more space in the cache.

To set the KV cache size, use the `-c, --context` parameter. For example, for 32 parallel streams that are expected to generate a maximum of 128 tokens each (i.e. `-n 128`), you would need to set `-c 4096` (i.e. 32*128). If continuous batching is enabled, you would need some extra KV space to deal with fragmentation of the cache. In the example above, we conveniently set the context size to 8192 to guarantee that there will be no issues.

Also, when computing the KV cache size you should also take into account the size of the system prompt (if any).

We can run a similar test on M2 Ultra. This time, since we have practically infinite VRAM, we can afford to run 128 parallel streams instead of just 32. And we can also use the F16 model instead of the quantum Q8_0. In this case, continuous batching is not needed, since we will be processing all of the 128 requests at once:

```bash
./bin/parallel -m ../models/llama-7b-v2/ggml-model-f16.gguf -n 128 -t 1 -ngl 100 -c 16384 -b 512 -s 1 -np 128 -ns 128
```

https://github.com/ggerganov/llama.cpp/assets/1991296/5768b011-e227-4297-98ad-64f23badd0c6

*(video is speed-up to fit in Github 10MB limit)*

### Implementation details

- KV cache as ring-buffer

  This is one of the major changes. Previously, the `n_past` number indicated what part of the KV cache to use in the computation. The assumption was that all cached data would be stored in the cache starting from the first cell forward and that this data belongs to one sequence only. We now want to store information from multiple sequences in the KV cache and the data from each sequence can be located in arbitrary cells. To achieve that, the KV cache now stores information about the sequences to which the cache data belongs and the corresponding position in these sequences:

  https://github.com/ggerganov/llama.cpp/blob/b377bf2266fcc7e98bef0ac2b8318b8b7c523947/llama.cpp#L1009-L1018

  With this change, we now have to attend to the entire cache when computing the self attention, because we don't know where the data for the sequences being processed is located. So the naive solution is when we build the graph to set `n_kv = n_ctx`. However, there is a simple heuristic that can be introduced in order to restore to most extend the behavior of `n_past` where we attend to just the first part of the cache:

  https://github.com/ggerganov/llama.cpp/blob/b377bf2266fcc7e98bef0ac2b8318b8b7c523947/llama.cpp#L2620-L2624

  Here, `kv_self.n` contains the index of the last "occupied" cell in the cache. I.e. we know that all cells beyond `kv_self.n` are currently empty, so no need to attend to them. We naively compute it each time we build a graph:

  https://github.com/ggerganov/llama.cpp/blob/b377bf2266fcc7e98bef0ac2b8318b8b7c523947/llama.cpp#L4101-L4107

  This heuristic is easy to implement and gives a measurable improvement, allowing to keep the performance the same as on `master` for single-sequence generations. It stops having any effect once the KV cache has been filled. In theory, we can improve this even further by computing `n_kv_0` and `n_kv_1` - the KV range that contains the cached tokens for all currently decoded sequences. We just need to find an elegant way to incorporate this information in the KV view offsets. (i.e. `n_kv = n_kv_1 - n_kv_0`, etc.).

### Workplan

- [x] verify that RoPE is "additive" (`tests/test-rope.cpp`)
- [x] update `ggml_rope` to take a tensor with positions instead of just `n_past`
  - [x] CPU
  - [x] Metal
    - [x] add `rope_f16` kernel
  - [x] CUDA
    - [x] update `rope_f32` kernel to work with vector of positions
    - [x] add `rope_f16` kernel (needed for K-cache shifts)
- [x] use `ggml_add` for applying a custom `-INF` mask to a tensor instead of the existing `ggml_diag_mask_inf`
  - [x] CPU
  - [x] Metal
    - [x] kernel `add` support broadcast across dims 1,2,3
  - [x] CUDA
- [x] update the graph to utilize the new `ggml` API for RoPE and mask
- [x] extend the KV cache to store position and sequence ID information
- [x] extend `llama.h` API for passing multi-sequence data
- [x] add `llama_kv_shift()` for "forgetting" old tokens, re-roping (if needed) the KV cache and compressing it
- [x] add example for generating N completions in parallel for a given prompt

### TODOs

- ~fix MPI~ will remain for a future PR
- [ ] deprecate `ggml_alibi()`, replace with `ggml_add()` (similar to `ggml_diag_mask_inf`, Baichuan 13B)
- [x] fix KV cache fragmentation with continuous batching

### Performance

I decided to disable the concurrency optimization since it prevents from modifying the graph topology. With the changes in this PR, we now need to sometimes add extra nodes for K-cache shift, so it is no longer compatible. The result is ~8% slower prompt processing but ~5% faster text generations with Metal.

M2 Ultra

| model                          |       size |    th  | test       |    master    t/s |      PR      t/s | speedup |
| ------------------------------ | ---------: | -----: | ---------- | ---------------: | ---------------: | ------: |
| LLaMA 7B mostly F16            |  12.55 GiB |      4 | pp 512     |   1490.37 ± 1.29 |   1381.20 ± 1.64 |  0.927  |
| LLaMA 7B mostly Q8_0           |   6.67 GiB |      4 | pp 512     |   1326.17 ± 0.49 |   1231.50 ± 0.99 |  0.929  |
| LLaMA 7B mostly Q4_0           |   3.56 GiB |      4 | pp 512     |   1355.75 ± 0.30 |   1258.24 ± 0.77 |  0.928  |
| LLaMA 7B mostly Q4_1           |   3.95 GiB |      4 | pp 512     |   1351.95 ± 0.58 |   1256.14 ± 0.97 |  0.929  |
| LLaMA 7B mostly Q6_K           |   5.15 GiB |      4 | pp 512     |   1106.49 ± 0.45 |   1033.03 ± 0.82 |  0.934  |
| LLaMA 7B mostly Q5_K - Medium  |   4.45 GiB |      4 | pp 512     |   1103.09 ± 0.65 |   1028.90 ± 1.40 |  0.933  |
| LLaMA 7B mostly Q5_K - Small   |   4.33 GiB |      4 | pp 512     |   1102.25 ± 0.33 |   1027.88 ± 0.83 |  0.933  |
| LLaMA 7B mostly Q4_K - Medium  |   3.80 GiB |      4 | pp 512     |   1168.98 ± 0.60 |   1089.69 ± 1.08 |  0.932  |
| LLaMA 7B mostly Q4_K - Small   |   3.59 GiB |      4 | pp 512     |   1178.79 ± 0.72 |   1097.91 ± 0.56 |  0.931  |
| LLaMA 7B mostly Q3_K - Medium  |   3.07 GiB |      4 | pp 512     |   1142.54 ± 0.65 |   1064.15 ± 0.39 |  0.931  |
| LLaMA 7B mostly Q3_K - Small   |   2.75 GiB |      4 | pp 512     |   1119.60 ± 0.62 |   1044.46 ± 0.53 |  0.933  |
| LLaMA 7B mostly F16            |  12.55 GiB |      4 | tg 128     |     40.94 ± 0.04 |     41.39 ± 0.04 |  1.011  |
| LLaMA 7B mostly Q8_0           |   6.67 GiB |      4 | tg 128     |     64.77 ± 0.05 |     67.84 ± 0.03 |  1.047  |
| LLaMA 7B mostly Q4_0           |   3.56 GiB |      4 | tg 128     |     91.14 ± 0.10 |     96.55 ± 0.09 |  1.059  |
| LLaMA 7B mostly Q4_1           |   3.95 GiB |      4 | tg 128     |     85.97 ± 0.10 |     89.57 ± 0.10 |  1.042  |
| LLaMA 7B mostly Q6_K           |   5.15 GiB |      4 | tg 128     |     71.44 ± 0.04 |     74.77 ± 0.06 |  1.047  |
| LLaMA 7B mostly Q5_K - Medium  |   4.45 GiB |      4 | tg 128     |     72.56 ± 0.05 |     75.44 ± 0.06 |  1.040  |
| LLaMA 7B mostly Q5_K - Small   |   4.33 GiB |      4 | tg 128     |     74.00 ± 0.06 |     76.96 ± 0.10 |  1.040  |
| LLaMA 7B mostly Q4_K - Medium  |   3.80 GiB |      4 | tg 128     |     83.71 ± 0.16 |     87.92 ± 0.12 |  1.050  |
| LLaMA 7B mostly Q4_K - Small   |   3.59 GiB |      4 | tg 128     |     87.05 ± 0.08 |     91.49 ± 0.08 |  1.051  |
| LLaMA 7B mostly Q3_K - Medium  |   3.07 GiB |      4 | tg 128     |     83.47 ± 0.14 |     87.62 ± 0.10 |  1.050  |
| LLaMA 7B mostly Q3_K - Small   |   2.75 GiB |      4 | tg 128     |     85.14 ± 0.14 |     88.23 ± 0.08 |  1.036  |

build: 897cacc (1272)

ggerganov committed:
tests : verify that RoPE is "additive"

ggerganov committed:
llama : replace ggml_diag_mask_inf with ggml_add (custom -inf mask)

ggerganov committed:
ggml : ggml_rope now takes a vector with positions instead of n_past

ggerganov commented on examples/train-text-from-scratch/train-text-from-scratch.cpp:
```diff
@@ -679,15 +679,23 @@ struct ggml_tensor * llama_build_train_graphs(
+            data[i] = n_past + i;
+        }
+    }
+
```
@xaedes I'm changing the API of `ggml_rope` to take an entire vector with positions instead of `n_past`. I have a small concern about this particular change in `train-text-from-scratch` and cannot test it atm. I'm not sure if the allocator won't make some intermediate results to overwrite the data of `KQ_pos` at some point.

In other places, we fix this using `ggml_allocr_alloc()`:

https://github.com/ggerganov/llama.cpp/blob/1fb033fd85f8125d2830bbfe6d384be3baa17ae8/llama.cpp#L2431-L2439

But wasn't sure if it's applicable here.

xaedes replied:
During training (finetune and train-text-from-scratch) n_past is always zero, so I guess KQ_pos would always be empty. 

> not sure if the allocator won't make some intermediate results to overwrite the data

To avoid deallocation of certain tensors T I added a temporary scale_inplace(T, 1.0f) operation at the end of the computation graph before giving it to the allocator. With this the allocator cannot deallocate T before the original end of the graph. Those temporary operations are removed from the graph after allocations are done.
For example here: https://github.com/ggerganov/llama.cpp/blob/5ce74ee4613c06bf3391c72d7115d10726200bff/examples/finetune/finetune.cpp#L768

ggerganov replied:
Hah, clever solution :) I added a scale op for `KQ_pos` to be safe.

> During training (finetune and train-text-from-scratch) n_past is always zero, so I guess KQ_pos would always be empty.

Btw, when `n_past == 0`, the `KQ_pos` tensor would have values `0, 1, 2, 3, ...` (i.e. `n_past + i`).

Cebtenzzre commented:
See my comment at https://github.com/ggerganov/llama.cpp/issues/2060#issuecomment-1722535693. We should consider implementing non-RoPEd KV cache in the future. I hope these changes wouldn't make that more difficult.

ggerganov commented:
@Cebtenzzre Shouldn't be more difficult than before this change. The main issue is updating `ggml_rope` to work with the transposed `V` data and potentially there is a concern about a performance hit due to having to RoPE the entire cache instead of just the new tokens. I'll probably give it a try within the scope of this PR, but first want to quickly finalize this version and make sure this entire approach is viable

ggerganov committed:
metal : add rope_f16 kernel + optimize cpy kernels

ggerganov committed:
llama : unified KV cache + batch inference API

ggerganov committed:
Merge branch 'master' into custom-attention-mask

ggerganov committed:
llama : add new llama_decode() API that works with llama_batch

ggerganov committed:
llama : add cell_max heuristic for more efficient kv_cache

ggerganov committed:
llama : extend llama_kv_cache API

ggerganov committed:
llama : more robust cell_max heuristic + wip shift

ggerganov committed:
metal : disable concurrency optimization

ggerganov committed:
llama : add llama_kv_cache_shift_seq + no more context swaps

ggerganov committed:
llama : apply K-cache roping for Falcon and Baichuan

ggerganov committed:
speculative : fix KV cache management

JohannesGaessler commented:
>The main issue is a concern about a performance hit due to having to RoPE the entire cache instead of just the new tokens.

I don't think this would be too bad. The main issue that would arise from RoPeing the KV cache on-the-fly would be that trigonometric functions are relatively expensive in terms of compute. For large batches you are compute bound but you still need to compute RoPe only once per eval so the compute *per token* should still be acceptable. Conversely, for small batches you are I/O bound so RoPEing the KV cache on-the-fly after loading it should not make a difference.

ggerganov committed:
parallel : example for serving multiple users in parallel

ggerganov committed:
parallel : disable hot-plug to avoid cache fragmentation

ggerganov committed:
fixes : speculative KV cache + llama worst-case graph

ggerganov commented:
I'm still thinking about the non-RoPEd K cache, but I guess I'm more leaning towards the current solution. I updated the graph to perform a "shift" of the K-cache when RoPE is involved via an optional call to `ggml_rope`:

https://github.com/ggerganov/llama.cpp/blob/897caccdf4248a3500f3b10a355b5eb487682c2d/llama.cpp#L2750-L2762

Since `ggml_rope` now takes a vector of positions (actually, deltas of the position) instead of just `n_past`, we can apply custom shifts to the entire KV cache or portions of it via the new `llama_kv_cache_shift_seq()`. This integrates well with the unified KV cache idea also implemented here, where we tore the cache from multiple independent sequences in the same buffer and utilize custom attention mask to attend to the correct data.

Overall, I think it works pretty well. There are some tricky things still happening and the devs have to be very careful. For example, the KV cache is not fully manually manages, so we have to be careful when to clear it and to clear it correctly.

I'll be looking the following days to polish the API and make things less error-prone. The branch should be ready for testing on CPU and Metal. It is lacking CUDA support, which I will try to implement, but would appreciate any help with it.

ggerganov committed:
llama : extend batch API to select which logits to output

slaren commented:
The buffer size calculation for falcon 7B is broken, it fails with `perplexity`:
```
perplexity: calculating perplexity over 12 chunks, batch_size=512
ggml_allocr_alloc: not enough space in the buffer (needed 133169152, largest block available 132118544)
GGML_ASSERT: ggml-alloc.c:173: !"not enough space in the buffer"
```

ggerganov committed:
llama : fix worst case graph build

ggerganov commented:
> The buffer size calculation for falcon 7B is broken,

Should be fixed now. I moved the K-cache shifting outside the layer loop. My guess is that when it is inside the loop, the graph that does rope shift actually ends up requiring less memory than the graph that does not require rope shift, which is counter-intuitive for determining the worst-case graph. But I'm not 100% this is the explanation for the failure, but after moving it, it seems to work

Will look now into merging the CUDA version

slaren committed:
ggml-cuda : update rope implementation for parallel decoding (#3254)

* ggml-cuda : update rope implementation for parallel decoding

* better solution for p0 computation

* fix rope

* simpler rope implementation

---------

Co-authored-by: Georgi Gerganov <ggerganov@gmail.com>

ggerganov committed:
make : add parallel to build + fix static functions in llama.cpp

ggerganov committed:
simple : fix token counting

ggerganov committed:
parallel : various improvements

ggerganov committed:
llama : fix cell_max logic + rename functions

ggerganov committed:
parallel : try smaller batches when the KV cache is fragmented

ggerganov committed:
parallel : fix sequence termination criteria

ggerganov committed:
llama : silence errors KV cache errors

ggerganov committed:
parallel : remove new line from prompt

slaren commented:
Currently this is crashing on generation with CUDA and CPU, and I think this is because sometimes `n_kv` is zero.

ggerganov commented:
What command do you use when it happens?

slaren commented:
I can reproduce it with just `./main -m models/7B/ggml-model-q4_0.gguf`. On CPU it fails with `GGML_ASSERT: ggml.c:11261: nb0 <= nb1`, with CUDA it's a different error because it tries to launch a kernel with 0 threads, because `KQ` has 0 elements.

ggerganov committed:
parallel : process system prompt once + configurable paramters + llama API

ggerganov committed:
parallel : remove question with short answers

ggerganov committed:
parallel : count cache misses

ggerganov committed:
parallel : print misses on each request

netrunnereve commented:
> I can reproduce it with just `./main -m models/7B/ggml-model-q4_0.gguf`. On CPU it fails with `GGML_ASSERT: ggml.c:11261: nb0 <= nb1`, with CUDA it's a different error because it tries to launch a kernel with 0 threads, because `KQ` has 0 elements.

I'm seeing the exact same crash on the latest commit with a x86 CPU.

ggerganov committed:
parallel : minor

ggerganov committed:
llama : fix n_kv to never become 0

ggerganov committed:
parallel : rename hot-plug to continuous-batching

ggerganov committed:
llama : improve llama_batch API + simplify parallel example

ggerganov committed:
simple : add parallel decoding support

ggerganov commented:
I've added a section "Implementation details" that currently contains some more info about the new KV cache and also explains the meaning of `n_kv`. I plan to extend this section with more info - let me know if there is anything particular of interest. I've also extended the `simple` example to demonstrate usage of the new `llama_batch` API for parallel decoding

ggerganov committed:
simple : improve comments + free batch

slaren committed:
ggml-cuda : add rope f16, restore performance with parallel decoding (#3272)

* ggml-cuda : add rope f16, restore performance

* offload KQ_mask with all models

* fix rope shift

---------

Co-authored-by: Georgi Gerganov <ggerganov@gmail.com>

ggerganov committed:
llama : disable MPI for now

ggml-ci

ggerganov committed:
train : make KQ_pos memory buffer permanent via dummy scale op

slaren committed:
ggml : revert change to ggml_cpy, add ggml_cont_Nd instead (#3275)

ggml-ci

ggerganov commented:
This is ready for thorough review. I will be travelling for the next few days, so the plan is to merge this PR some time after Tue 26 Sep.

Before merging, I would like to improve the API some more based on usage feedback and add more comments to the API and the examples. Overall, I'm happy with the functionality and it should enable all kinds of techniques, mainly related but not limited to speculative decoding.

Need to fix Baichuan 13B alibi operator, but I don't have the model to test with. @jameswu2014 or anyone else - would be great if you attempt to fix this before merging (see TODOs and let me know if something is not clear). Should be easy change, but need to setup testing with the actual model.

Also, have to write a thorough summary of the API changes so that it is easier for third-party projects to adapt. Basically, the old API should be transparently supported, but it is marked as DEPRECATED and it's better for everyone to migrate to the new `llama_decode` API.

While doing the changes in `llama.cpp` I think it might be time to unify the `llm_build_*` functions for graph builds to reduce code duplication. Will probably create a refactoring plan after merging this.

ggerganov committed:
parallel : fix bug (extra BOS) + smaller token_prev array

AutonomicPerfectionist commented:
> Unfortunately, I don't see an easy workaround for the MPI build - we need to sync the pos array, not just n_past. It's probably not something very difficult, but prefer to deal with this later

I'm currently doing some work with the MPI build for a master's class that involves a decent refactoring to support heterogenous clusters (per node thread and layer count), so I can include syncing the pos array with my other changes. Hoping to have a draft PR up this weekend for comment

ggerganov commented:
@AutonomicPerfectionist That would be great! I haven't thought too much into the details, but I think if you sync everything from `llama_batch` except `token` and `embd` (i.e. `pos`, `seq_id` and `logits`) it should work. But I could be missing something. Ah, and the other thing is syncing the KV cache ops, like `llama_kv_cache_seq_rm`, but this would be needed just for more complex use cases (i.e. parallel decoding, rewinding back to a previous state, etc.).

ggerganov committed:
parallel : fix cases where the input prompts can overflow the batch

ggerganov committed:
parallel : add disabled experimental batch chunking in powers of two

kiratp commented:
I know PagedAttention was tried and dropped due to no performance change but it may be worth verifying if there is a performance benefit here given the amortized cost of RAM reads with this PR. 

Really excited to see this come together!

ggerganov committed:
llama : llama.h formatting + comments

ggerganov committed:
simple : add README.md

pudepiedj commented:
I have run this successfully on both a **q4_0** and **q8_0** quantized version of ggml-model-f16.gguf on an M2 MAX (38 core) with 32GB RAM with 128 clients.
```
% ./bin/parallel -m ../models/llama-2-7b/ggml-model-q8_0.gguf -n 128 -t 1 -ngl 100 -c 8192 -b 512 -s 1 -np 32 -ns 128 -cb
```
The q8_0 runs in about 57.3s:
```
main: clearing the KV cache

Total prompt tokens:   1848, speed: 32.25 t/s
Total gen tokens:      8691, speed: 151.68 t/s
Total speed (AVG):           speed: 183.94 t/s
Cache misses:             0

llama_print_timings:        load time = 10096.03 ms
llama_print_timings:      sample time =  5680.08 ms /  8819 runs   (    0.64 ms per token,  1552.62 tokens per second)
llama_print_timings: prompt eval time = 49950.62 ms / 10823 tokens (    4.62 ms per token,   216.67 tokens per second)
llama_print_timings:        eval time =   811.50 ms /    21 runs   (   38.64 ms per token,    25.88 tokens per second)
llama_print_timings:       total time = 57298.74 ms
ggml_metal_free: deallocating

```
The q4_0 in 53.2s
```
Total prompt tokens:   1848, speed: 34.70 t/s
Total gen tokens:      8221, speed: 154.37 t/s
Total speed (AVG):           speed: 189.08 t/s
Cache misses:             0

llama_print_timings:        load time =   704.10 ms
llama_print_timings:      sample time =  5347.68 ms /  8349 runs   (    0.64 ms per token,  1561.24 tokens per second)
llama_print_timings: prompt eval time = 46528.23 ms / 10354 tokens (    4.49 ms per token,   222.53 tokens per second)
llama_print_timings:        eval time =   588.89 ms /    20 runs   (   29.44 ms per token,    33.96 tokens per second)
llama_print_timings:       total time = 53254.37 ms
ggml_metal_free: deallocating
```
I can't run the 128 client version of the f16 using 16384
```
% ./bin/parallel -m ../models/llama-2-7b/ggml-model-f16.gguf -n 128 -t 1 -ngl 100 -c 16384 -b 512 -s 1 -np 128 -ns 128 
```
```
ggml_metal_init: hasUnifiedMemory              = true
ggml_metal_init: recommendedMaxWorkingSetSize  = 21845.34 MB
ggml_metal_init: maxTransferRate               = built-in GPU
llama_new_context_with_model: compute buffer total size = 1081.47 MB
llama_new_context_with_model: max tensor size =   250.00 MB
ggml_metal_add_buffer: allocated 'data            ' buffer, size = 12853.73 MB, (12854.23 / 21845.34)
ggml_metal_add_buffer: allocated 'eval            ' buffer, size =     1.48 MB, (12855.72 / 21845.34)
ggml_metal_add_buffer: allocated 'kv              ' buffer, size =  8194.00 MB, (21049.72 / 21845.34)
ggml_metal_add_buffer: allocated 'alloc           ' buffer, size =  1080.02 MB, (22129.73 / 21845.34), warning: current allocated size is greater than the recommended max working set size
ggml_metal_graph_compute: command buffer 0 failed with status 5
GGML_ASSERT: /Users/edsilm2/llama.cpp/ggml-metal.m:1337: false
zsh: abort      ./bin/parallel -m ../models/llama-2-7b/ggml-model-f16.gguf -n 64 -t 1 -ngl 10
```
but the unquantized f16 runs for 128 clients using 8192 in 75.9s.
```
% ./bin/parallel -m ../models/llama-2-7b/ggml-model-f16.gguf -n 128 -t 1 -ngl 100 -c 8192 -b 512 -s 1 -np 128 -ns 128 
```
```
Total prompt tokens:   1848, speed: 24.34 t/s
Total gen tokens:      8436, speed: 111.09 t/s
Total speed (AVG):           speed: 135.43 t/s
Cache misses:           273



llama_print_timings:        load time = 18456.44 ms
llama_print_timings:      sample time =  5611.87 ms /  8564 runs   (    0.66 ms per token,  1526.05 tokens per second)
llama_print_timings: prompt eval time = 69006.75 ms / 10582 tokens (    6.52 ms per token,   153.35 tokens per second)
llama_print_timings:        eval time =   470.16 ms /     7 runs   (   67.17 ms per token,    14.89 tokens per second)
llama_print_timings:       total time = 75938.57 ms

```
The q8_0 runs in 16394 in 63.6s:
```
Total prompt tokens:   1848, speed: 29.06 t/s
Total gen tokens:      8597, speed: 135.18 t/s
Total speed (AVG):           speed: 164.24 t/s
Cache misses:             0

llama_print_timings:        load time = 10167.61 ms
llama_print_timings:      sample time =  5621.54 ms /  8725 runs   (    0.64 ms per token,  1552.07 tokens per second)
llama_print_timings: prompt eval time = 57092.29 ms / 10749 tokens (    5.31 ms per token,   188.27 tokens per second)
llama_print_timings:        eval time =    54.07 ms /     1 runs   (   54.07 ms per token,    18.50 tokens per second)
llama_print_timings:       total time = 63597.80 ms
ggml_metal_free: deallocating

```
And the q4_0 in 58.4s
```
Total prompt tokens:   1848, speed: 31.66 t/s
Total gen tokens:      8259, speed: 141.50 t/s
Total speed (AVG):           speed: 173.17 t/s
Cache misses:             0

llama_print_timings:        load time =  5610.70 ms
llama_print_timings:      sample time =  5512.67 ms /  8387 runs   (    0.66 ms per token,  1521.40 tokens per second)
llama_print_timings: prompt eval time = 52044.21 ms / 10412 tokens (    5.00 ms per token,   200.06 tokens per second)
llama_print_timings:        eval time =     0.00 ms /     1 runs   (    0.00 ms per token,      inf tokens per second)
llama_print_timings:       total time = 58366.43 ms
ggml_metal_free: deallocating

```
I hope this helps in some small way.

nytopop commented:
Performance on cuda is a bit odd - setting `-np` higher than 1 causes a significant reduction in tokens/s. Naively, I would expect throughput to increase ~linearly from 1 until it becomes compute bound. Instead there's a sharp downturn at 2 and from there it behaves reasonably.

I don't see this happen on cpu, there `tokens / s` increases smoothly until compute bound is reached.

Graphs to show what I'm talking about (x axis is `np`, y is `tokens/s`, and red / blue / green are `prompt` / `gen` / `average` reported at the end of each run):

* HIPBLAS, amd instinct mi60 (codellama-7b.Q8_0.gguf)

![2023-09-24-015524_938x626_scrot](https://github.com/ggerganov/llama.cpp/assets/7799664/e7c11ac8-0fa5-4a2e-9c00-c0ecae5d80d1)

* CUBLAS, nvidia gtx 1080 (codellama-7b.Q4_0.gguf)

![2023-09-24-015539_761x608_scrot](https://github.com/ggerganov/llama.cpp/assets/7799664/168b94cb-58a0-44de-905f-018f0e99f64f)

* cpu, i7-7820x (codellama-7b.Q4_0.gguf)

![2023-09-24-024155_1040x865_scrot](https://github.com/ggerganov/llama.cpp/assets/7799664/11f68d10-ac0f-4d8e-b907-ddf2cf3dfa28)

pudepiedj commented:
@ggerganov: Running the ./bin/parallel code without setting `-n` or by setting it to `-1 (default)` or `-2` with **f16**, **q4_0** or **q8_0** generates an uncaught exception runtime error on M2 MAX:
```
libc++abi: terminating due to uncaught exception of type std::length_error: vector
zsh: abort      ./bin/parallel -m ../models/llama-2-7b/ggml-model-q8_0.gguf -n -2 -t 8 -ngl 
```

JohannesGaessler commented:
@nytopop That is to be expected. There are 2 different CUDA implementations either for a batch size of 1 or a batch size >1. The implementation for >1 was optimized for large batches and has poor performance small batches.

pudepiedj commented:
I have run the first and third of the Demo examples in the `origin/custom-attention-mask` branch successfully on an M2 MAX 32GB MacOS 13.5.1, which suggests that the install is working, but I can't make the second **./simple** example run past the kernel definitions when it throws this **computeFunction must not be nil** error:
```
ggml_metal_init: loaded kernel_mul_mm_q4_K_f32                0x12a11b2d0 | th_max =  768 | th_width =   32
ggml_metal_init: loaded kernel_mul_mm_q5_K_f32                0x12a11bb00 | th_max =  768 | th_width =   32
ggml_metal_init: loaded kernel_mul_mm_q6_K_f32                0x12a11c590 | th_max =  768 | th_width =   3
-[MTLComputePipelineDescriptorInternal setComputeFunction:withType:]:692: failed assertion `computeFunction must not be nil.'
zsh: abort      ./simple ./models/llama-2-7b/ggml-model-f16.gguf "Hello! My name is" 8
```
It does this on **f16**, **q4_0** and **q8_0** with parameters `1, 4, 8`. It is of course possible that this is a local problem but I've run **cmake** from scratch twice and checked all the versions including **ggml-metal.metal** to look at the kernel definitions, but I can't track the problem down.

slaren commented:
I think it would be good to keep the current `simple` example as is (updated to the new API), and move the new parallel simple example to a new example. While it is a small example, the logic is a lot more complicated than generating a single stream, and many users will never need to do parallel evaluations.

Additionally, I think we should remove all the dependencies with `common.h/cpp` from these examples, users should be able to use these examples as a starting point without any dependencies.

slaren commented on llama.cpp:
```diff
@@ -6951,64 +7375,112 @@ bool llama_save_session_file(struct llama_context * ctx, const char * path_sessi
-        ctx->t_load_us = ggml_time_us() - ctx->t_start_us;
-        ctx->has_evaluated_once = true;
+    const int ret = llama_decode_internal(*ctx, llama_batch_get_one(tokens, n_tokens, n_past, 0), n_threads);
+    if (ret != 0) {
+        if (ret < 0) {
+            LLAMA_LOG_ERROR("%s: failed to decode, ret = %d\n", __func__, ret);
+        }
+
+        return ret;
     }
```
```suggestion
    if (ret < 0) {
        LLAMA_LOG_ERROR("%s: failed to decode, ret = %d\n", __func__, ret);
    }
```

ggerganov committed:
llama : fix kv cache heuristic when context is less than 32

pudepiedj commented:
See below for at least an interim solution.

> I have run the first and third of the Demo examples in the `origin/custom-attention-mask` branch successfully on an M2 MAX 32GB MacOS 13.5.1, which suggests that the install is working, but I can't make the second **./simple** example run past the kernel definitions when it throws this **computeFunction must not be nil** error:
> 
> ```
> ggml_metal_init: loaded kernel_mul_mm_q4_K_f32                0x12a11b2d0 | th_max =  768 | th_width =   32
> ggml_metal_init: loaded kernel_mul_mm_q5_K_f32                0x12a11bb00 | th_max =  768 | th_width =   32
> ggml_metal_init: loaded kernel_mul_mm_q6_K_f32                0x12a11c590 | th_max =  768 | th_width =   3
> -[MTLComputePipelineDescriptorInternal setComputeFunction:withType:]:692: failed assertion `computeFunction must not be nil.'
> zsh: abort      ./simple ./models/llama-2-7b/ggml-model-f16.gguf "Hello! My name is" 8
> ```
> 
> It does this on **f16**, **q4_0** and **q8_0** with parameters `1, 4, 8`. It is of course possible that this is a local problem but I've run **cmake** from scratch twice and checked all the versions including **ggml-metal.metal** to look at the kernel definitions, but I can't track the problem down.

I seem to have resolved this. During the `cmake` process the **simple.cpp** source was being compiled using a different version of **ggml-metal.m** as a result of it being referenced in the **sync-ggml.sh** script in `llama.cpp/scripts`.
```
cp -rpv ../ggml/src/ggml-metal.m          ./ggml-metal.m
```

Once I edited that to point to the version inside `llama.cpp` that was working with the other two demo programs, it worked straight away.

I have no idea how or why this came about (and why it didn't affect the first and third of the demo programs) because I have never edited the `sync-ggml.sh` file, and the version in the `master` repo points to the version in the `ggml/` repo created by cloning the `ggerganov/ggml` repo as well. Is it possible that there is something wrong with the `ggml-model-m` file in that directory; at least, `./simple` didn't work when compiled with it on my M2 MAX as reported above.

In the working version, `./simple` is compiled into `build/bin' so running it is more like `./bin/parallel`.
```
./bin/simple ../models/llama-2-7b/ggml-model-f16.gguf "Hello my name is" 8
```
@ggerganov I don't understand the intricacies of all this sufficiently to know whether this was just a result of my stupidity or indicative of another issue, so I simply report it as I found it for what it is worth.

Cebtenzzre commented:
> seem to have resolved this. During the `cmake` process the **simple.cpp** source was being compiled using a different version of **ggml-metal.m** as a result of it being referenced in the **sync-ggml.sh** script in `llama.cpp/scripts`.

Why are you running sync-ggml.sh?

ggerganov committed:
Merge branch 'master' into custom-attention-mask

ggerganov committed:
parallel : fix crash when `-n -1`

ggerganov committed:
llama : simplify returns if/else branches

ggerganov committed:
metal : use mm kernels for batch size > 2

pudepiedj commented:
> > seem to have resolved this. During the `cmake` process the **simple.cpp** source was being compiled using a different version of **ggml-metal.m** as a result of it being referenced in the **sync-ggml.sh** script in `llama.cpp/scripts`.
> 
> Why are you running sync-ggml.sh?

Hmmm ... good question. I didn't think I _was_ running it and it appears that I probably _wasn't_, but the only thing I changed between **simple** compiling but failing to run with one version of **ggml-metal.m** and compiling with the other one that made it work was the line above in **sync-ggml.sh**. So I assumed that had been the problem and that something was running **sync-ggml .sh** somewhere during the compile process. I have now reversed my changes and **./simple** still works and still compiles with the same **ggml-model.m** version as the other two demos, so obviously a red herring. Sorry to waste your time. Investigations continue ...

ggerganov committed:
examples : utilize new llama_get_logits_ith()

ggerganov committed:
examples : add example for batched decoding

slaren commented:
https://github.com/ggerganov/llama.cpp/pull/3348 has fixes for some issues that I found with the simple example (now called batched). I don't think it can be merged anymore, but maybe the fixes can be added to the new example.

ggerganov committed:
examples : do not eval prompt 2 times (close #3348)

ggerganov commented:
This should be ready to merge - anything else before doing so?

slaren commented:
Does the server example need to be updated to erase unused parts of the KV cache? I am thinking that here, if `n_past` is reduced, that part of the KV cache should also be erased to avoid creating a leak in the KV cache:
https://github.com/ggerganov/llama.cpp/blob/2b8830af7153a75fba5a899a8b331389025d0d03/examples/server/server.cpp#L382-L383

ggerganov committed:
server : clear the KV cache beyond n_past before llama_decode

ggerganov committed:
server : avoid context swaps by shifting the KV cache

pudepiedj commented:
Worth adding a reference to the **total sequence** since the sequence is not ordered and maybe some **colour coding**?
```
Client   0, seq  123/128, prompt   22 t, response  106 t, time 11.46 s, speed 11.17 t/s, cache miss 0  

Input:    Are you familiar with the Special Theory of Relativity and can you explain it to me?
Response: The Special Theory of Relativity is a theory of physics that states that the speed of light in a vacuum is constant and independent of the observer. It also states that the laws of physics are the same for all observers, regardless of their relative motion. This means that time, space, and matter are affected by the velocity of the observer, and that the laws of physics are the same for all observers, regardless of their relative motion. To explain this in simple terms, imagine two people standing on opposite sides of a

main: clearing the KV cache

Total prompt tokens:   1848, speed: 35.35 t/s
Total gen tokens:      8300, speed: 158.76 t/s
Total speed (AVG):           speed: 194.11 t/s
Cache misses:             0

llama_print_timings:        load time =   352.25 ms
llama_print_timings:      sample time =  5563.68 ms /  8428 runs   (    0.66 ms per token,  1514.83 tokens per second)
llama_print_timings: prompt eval time = 45029.00 ms / 10425 tokens (    4.32 ms per token,   231.52 tokens per second)
llama_print_timings:        eval time =   865.89 ms /    28 runs   (   30.92 ms per token,    32.34 tokens per second)
llama_print_timings:       total time = 52281.72 ms
ggml_metal_free: deallocating
```
Easy to do:
```
LOG_TEE("\033[31mClient %3d, seq %4d/%3d, prompt %4d t, response %4d t, time %5.2f s, speed %5.2f t/s, cache miss %d \033[0m \n\nInput:    %s\nResponse: %s\n\n",
                            client.id, client.seq_id, n_seq, client.n_prompt, client.n_decoded,

```

henk717 commented:
> @nytopop That is to be expected. There are 2 different CUDA implementations either for a batch size of 1 or a batch size >1. The implementation for >1 was optimized for large batches and has poor performance small batches.

Unfortunately that makes it unfit for our own use case, multiple of our users wanted to run volunteer workers at the AI Horde. But demand fluctuates heavily and can be lower, so during the lower thread systems / lower throughput models this would harm our performance if used.

So for our hobbyist driven platform we seem to land right in the middle where performance is the worst.

ggerganov commented:
We should look into improving the performance for small batch size. This would be useful for the speculative decoding approaches that we are going to implement soon

pudepiedj commented:
I couldn't get the `./bin/parallel` external alternative-prompt file option to work using `-f file.txt` so I've inserted a few lines into `parallel.cpp` to make it run. I am not a `C/C++` specialist so it probably won't pass any style tests (code at the end), but it does run on my **M2 MAX 32GB** and **MacOS Sonoma 14.0** from
```
% ./build/bin/parallel -m ./models/llama-2-13b/ggml-model-q8_0.gguf -f "ParallelQuestions.txt" -n 128 -t 1 -c 8192 -s 4321 -ngl 100 -np 16 -ns 32 -cb
```
A plain text file like this below saved into `/build/ParallelQuestions.txt` is then read in generically by `common.cpp` (with no alterations to the code) and transferred to replace the default `k_prompts` inside `parallel.cpp`.
```
What do you know about Hobbits?
What is quantum field theory?
Why did the chicken cross the road?
Who is the president of the United States?
How do I run CMake on MacOS?
Do you agree that C++ is a really finicky language compared with Python3?
Is it a good idea to invest in technology?
Do you like Wagner's Ring?
Do you think this file input option is really neat?
What should we all do about climate change?
```
Producing (sample only) output
```
Now printing the k_prompts loaded from ParallelQuestions.txt that replace the default questions.

What do you know about Hobbits?
What is quantum field theory?
Why did the chicken cross the road?
Who is the president of the United States?
How do I run CMake on MacOS?
Do you agree that C++ is a really finicky language compared with Python3?
Is it a good idea to invest in technology?
Do you like Wagner's Ring?
Do you think this file input option is really neat?
What should we all do about climate change?


main: Simulating parallel requests from clients:
main: Evaluating the system prompt ...

Processing requests ...

main: clearing the KV cache
Client   0, seq    0, started decoding ...
Client   1, seq    1, started decoding ...
Client   2, seq    2, started decoding ...
Client   3, seq    3, started decoding ...
Client   4, seq    4, started decoding ...
Client   5, seq    5, started decoding ...
Client   6, seq    6, started decoding ...
Client   7, seq    7, started decoding ...
Client   8, seq    8, started decoding ...
Client   9, seq    9, started decoding ...
Client  10, seq   10, started decoding ...
Client  11, seq   11, started decoding ...
Client  12, seq   12, started decoding ...
Client  13, seq   13, started decoding ...
Client  14, seq   14, started decoding ...
Client  15, seq   15, started decoding ...
Client   7, seq    7/ 32, prompt   14 t, response   11 t, time  2.62 s, speed  9.54 t/s, cache miss 0  
Input:    Who is the president of the United States?
Response: The current President of the United States is Donald Trump.

Client  14, seq   14/ 32, prompt   14 t, response   11 t, time  2.63 s, speed  9.52 t/s, cache miss 0  
Input:    Who is the president of the United States?
Response: The current President of the United States is Donald Trump.

Client   7, seq   16, started decoding ...
Client  14, seq   17, started decoding ...
Client   7, seq   16/ 32, prompt   14 t, response   13 t, time  2.39 s, speed 11.28 t/s, cache miss 0  
Input:    Why did the chicken cross the road?
Response: The chicken crossed the road to get to the other side.

Client   7, seq   18, started decoding ...
Client  12, seq   12/ 32, prompt   14 t, response   30 t, time  5.87 s, speed  7.49 t/s, cache miss 0  
Input:    Who is the president of the United States?
Response: The current president of the United States is Joe Biden. He was sworn into office on January 20, 2021.

Client  12, seq   19, started decoding ...
Client  10, seq   10/ 32, prompt   13 t, response   35 t, time  6.73 s, speed  7.13 t/s, cache miss 0  
Input:    What should we all do about climate change?
Response: We should all work together to reduce our carbon footprint. We can do this by reducing our energy consumption, recycling more, and using public transportation whenever possible.

Client  10, seq   20, started decoding ...
Client   1, seq    1/ 32, prompt   14 t, response   42 t, time  7.93 s, speed  7.06 t/s, cache miss 0  
Input:    Why did the chicken cross the road?
Response: The chicken crossed the road to get to the other side. It is a classic joke that has been around for many years and is often used as an example of a riddle or paradox.

Client   1, seq   21, started decoding ...
Client  14, seq   17/ 32, prompt   14 t, response   36 t, time  6.36 s, speed  7.86 t/s, cache miss 0  
Input:    Who is the president of the United States?
Response: The president of the United States is Joe Biden. He was elected in November 2020 and took office on January 20, 2021.

Client  14, seq   22, started decoding ...
Client   0, seq    0/ 32, prompt   16 t, response   49 t, time  9.17 s, speed  7.09 t/s, cache miss 0  
Input:    Do you think this file input option is really neat?
Response: Yes, I think this file input option is really neat. It allows you to easily upload files from your computer or other devices and view them in the browser. This can be very useful when sharing large files or working with multiple documents at once.

Client   0, seq   23, started decoding ...
Client  10, seq   20/ 32, prompt   13 t, response   14 t, time  2.62 s, speed 10.29 t/s, cache miss 0  
Input:    Do you like Wagner's Ring?
Response: I do not know what "Wagner's Ring" is.

Client  10, seq   24, started decoding ...
Client   9, seq    9/ 32, prompt   22 t, response   56 t, time 10.42 s, speed  7.49 t/s, cache miss 0  
Input:    Do you agree that C++ is a really finicky language compared with Python3?
Response: I cannot agree or disagree with that statement because it is subjective and depends on the person who is asking the question. Some people may find C++ to be finicky, while others may not have any problems with it. Ultimately, it comes down to personal preference.
```
The additional lines of code (does C++ not have a .split("\n") predefined function?):
```
// Define a split string function to ...
    std::vector<std::string> splitString(const std::string& input, char delimiter) {
        std::vector<std::string> tokens;
        std::istringstream stream(input);
        std::string token;
        while (std::getline(stream, token, delimiter)) {
            tokens.push_back(token);
        }
        return tokens;
    }
```
And then somewhere suitable inside `main()` and before `k_prompts` is used, in my case roughly at line 125:
```
    // load the prompts from an external file if there are any
    // these have been acquired by `common.cpp` and put into `params.prompt`
    // by parsing `-f ParallelQuestions.txt` in lines 150-164

    bool exitIf = false;
    if (params.prompt.empty()) {
        exitIf = true; // no new questions so skip the rest and carry on using defaults
    }
    if (exitIf) {
        // Output each line of the input params.prompts vector and load into k_prompts
        int index = 0;
        std::cout << "\nNow printing the original prompt file\n\n";

        std::vector<std::string> tokens = splitString(params.prompt, '\n');
        for (const auto& token : tokens) {
            k_prompts.resize(index + 1);
            k_prompts[index] = token;
            index++;
            std::cout << index << " token " << token << std::endl;
        }

        // Output each line of the updated k_prompts vector (diagnostic check non-functional)
        std::cout << "\n\033[33mNow printing the k_prompts loaded from ParallelQuestions.txt that replace the default questions.\033[0m\n\n";
        for (const auto& prompt : k_prompts) {
            std::cout << prompt << std::endl;
        }
    }
```

Mihaiii commented:
What does the "custom attention mask" mentioned in the title means?

I thought it means being able to emphasis some parts of the prompt similar with how [SD community does it](https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Features#attentionemphasis) (ex: "Do (not:1.5) do X"), but after browsing the code from this PR, I suspect I was wrong.

IridiumMaster commented:
Will this (very neat) addition be incorporated into the server example?

pudepiedj commented:
> Will this (very neat) addition be incorporated into the server example?

Thanks! I've opened a version of this as PR #3416. It was fun to test it using `./examples/jeopardy/questions.txt`.
```
./build/bin/parallel -m ./models/llama-2-13b/ggml-model-q8_0.gguf -f "./examples/jeopardy/questions.txt" -n 128 -t 4 -s 1 -ngl 100 -c 8192 -np 64 -ns 128 -cb
```
Sample output on M2 MAX 32GB:
```
Client  47, seq 124/128, prompt   44 t, response   84 t, time 30.12 s, speed  4.25 t/s, cache miss 0  
Input:    In a TV reunion over 40 years in the making, Dolly Parton appeared as an angel named Agnes in the final episode of what comedy in 2022?
Response: Dolly Parton appeared as an angel named Agnes in the final episode of "Grace and Frankie" in 2022. The comedy was created by Marta Kauffman and Howard J. Morris, and it starred Jane Fonda and Lily Tomlin as Grace and Frankie, two women who are brought together when their husbands reveal they have been having an affair for

Client  63, seq 127/128, prompt   34 t, response   94 t, time 28.76 s, speed  4.45 t/s, cache miss 0  
Input:    Which decorative items in fashion history get their name from their origin in the port city of Strasbourg, on the border of France & Germany?
Response: You are asking about the origin of the name of decorative items that get their name from the port city of Strasbourg. According to Wikipedia, the term "Strasbourg" was used to describe several different types of ornamental items that were popular in the 18th and 19th centuries. These items included furniture, ceramics, and textiles, all of which were made using a specific technique or style that originated in Stras

main: clearing the KV cache

RUN PARAMETERS as at Sun Oct  1 11:27:24 2023

main: n_parallel = 64, n_sequences = 128, cont_batching = 1, system tokens = 305
external prompt file (if any): ./examples/jeopardy/questions.txt

Total prompt tokens:   4857, speed: 52.86 t/s
Total gen tokens:      5396, speed: 58.73 t/s
Total speed (AVG):           speed: 111.60 t/s
Cache misses:             0


llama_print_timings:        load time = 19798.84 ms
llama_print_timings:      sample time =  3724.41 ms /  5524 runs   (    0.67 ms per token,  1483.19 tokens per second)
llama_print_timings: prompt eval time = 86529.40 ms / 10545 tokens (    8.21 ms per token,   121.87 tokens per second)
llama_print_timings:        eval time =  1044.25 ms /    13 runs   (   80.33 ms per token,    12.45 tokens per second)
llama_print_timings:       total time = 91881.77 ms
```

ggerganov commented:
> What does the "custom attention mask" mentioned in the title mean?

It means we can now do fine-grained control for each token to which other tokens it can attend to during the attention computation. This enables a large set of functionality that was not possible before due to supporting just the standard causal triangular attention mask.

cebtenzzre commented on llama.h:
```diff
@@ -289,48 +367,100 @@ extern "C" {
+    // NOTE: this is a helper function to facilitate transition to the new batch API - avoid using it
+    //
+    LLAMA_API struct llama_batch llama_batch_get_one(
+                  llama_token * tokens,
```
Why is this not `const llama_token * tokens`?

ggerganov replied:
Because `llama_batch.tokens` is not `const`. I initially tried to make it `const`, but it is too restrictive. Also, after adding `llama_batch_init()` we want to be able to mutate the newly allocated `llama_batch`, so no point in `const`

cebtenzzre replied:
So if I want to pass in a list of tokens from a const vector in C++, as I did before without issue, is the correct use of the API `const_cast<llama_token *>(tokens.data())`? Or do I need to copy the vector first?

ggerganov replied:
Ideally, you want to `llama_batch_init` a new object and use that all the time. Alternatively, un-const the C++ vector if that is an option. `const_cast` works, but I want to avoid it fully in this project, even in the examples, so better not use that.

MrJackSpade commented:
Is the intent of the cache shifting to allow for arbitrary shifting of data in any direction?

I've recently run into an issue with receiving garbage data from the model, and it appears to be related to shifting cells with a + delta. 

I know the immediate use case is to - shift to account for context overflow, however I was + shifting to insert data into the prompt, and that seems to be what's causing everything to break down.

I know its not a standard use case, but since I don't see anything in the code preventing use of a + delta, I assumed it was supported. 

I've been trying to dig in and see if I can figure out exactly what's going wrong, but I'm getting in a bit over my head in the code trying to chase it down.

GentlemanOfCulture commented:
I see the warning about using world info, but how exactly does context shifting work with world info? If a world entry is triggered and is added to the exact middle of the context, for example, then what portions of the prompt need re-evaluation? Is it just the world entry itself, or is it all context that follows the world entry?
